File: frontend/intuitiveapp/src/services/refeicaoService.ts
import api from './api'
import { extrairMensagemErro } from './authService'

export interface Refeicao {
  idRefeicao?: number
  data: string
  horario: string
  tipo: string
  descricao?: string
  nivelFome: number
  nivelSaciedade: number
  companhia?: string
  distracoes?: string
  emocoesAntes?: string
  emocoesDepois?: string
  usuario: {
    id: number
    nome?: string
    email?: string
  }
}

export const refeicaoService = {
  async cadastrar(refeicao: Omit<Refeicao, 'idRefeicao'>): Promise<Refeicao> {
    try {
      const response = await api.post('/refeicao', refeicao)
      return response.data
    } catch (error: any) {
      return Promise.reject({ message: extrairMensagemErro(error, 'Erro ao cadastrar refeição') })
    }
  },

  async consultarPorId(id: number): Promise<Refeicao> {
    try {
      const response = await api.get(`/refeicao/${id}`)
      return response.data
    } catch (error: any) {
      return Promise.reject({ message: extrairMensagemErro(error, 'Erro ao consultar refeição') })
    }
  },

  async getHistoricoMensal(usuarioId: number, mes: number, ano: number): Promise<Refeicao[]> {
    try {
      const response = await api.get('/refeicao/historico', { params: { usuarioId, mes, ano } })
      return response.data
    } catch (error: any) {
      return Promise.reject({ message: extrairMensagemErro(error, 'Erro ao carregar histórico') })
    }
  },

  async getMesesDisponiveis(usuarioId: number, ano: number) {
    try {
      const response = await api.get('/refeicao/meses-disponiveis', { params: { usuarioId, ano } })
      return response.data
    } catch (error: any) {
      return Promise.reject({ message: extrairMensagemErro(error, 'Erro ao carregar meses disponíveis') })
    }
  },

  async atualizar(id: number, refeicao: Refeicao): Promise<Refeicao> {
    try {
      const response = await api.put(`/refeicao/${id}`, refeicao)
      return response.data
    } catch (error: any) {
      return Promise.reject({ message: extrairMensagemErro(error, 'Erro ao atualizar refeição') })
    }
  },

  async excluir(idRefeicao: number): Promise<void> {
    try {
      await api.delete(`/refeicao/${idRefeicao}`)
    } catch (error: any) {
      return Promise.reject({ message: extrairMensagemErro(error, 'Erro ao excluir refeição') })
    }
  }
}

File: backend/src/main/java/com/intuitive/app/business/RefeicaoService.java
package com.intuitive.app.business;

import com.intuitive.app.DTO.RefeicaoDto;
import com.intuitive.app.DTO.UsuarioDto;
import com.intuitive.app.infrastructure.entitys.Refeicao;
import com.intuitive.app.infrastructure.entitys.Usuario;
import com.intuitive.app.infrastructure.repository.RefeicaoRepository;
import com.intuitive.app.infrastructure.repository.UsuarioRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class RefeicaoService {

    private final RefeicaoRepository repository;
    private final UsuarioRepository usuarioRepository;

    // ===================== Cadastro =====================
    public RefeicaoDto salvarRefeicao(RefeicaoDto dto) {
        validarCamposObrigatorios(dto);

        Usuario usuario = usuarioRepository.findById(dto.getUsuario().getId())
                .orElseThrow(() -> new RuntimeException("Usuário não encontrado"));

        Refeicao refeicao = toEntity(dto, usuario);
        Refeicao salvo = repository.save(refeicao);

        return toDTO(salvo);
    }

    // ===================== Consulta por ID =====================
    public RefeicaoDto consultarRefeicao(Integer idRefeicao) {
        Refeicao refeicao = repository.findById(idRefeicao)
                .orElseThrow(() -> new RuntimeException("Refeição não encontrada"));
        return toDTO(refeicao);
    }

    // ===================== Atualização =====================
    public RefeicaoDto atualizarRefeicao(Integer id, RefeicaoDto dto) {
        Refeicao refeicaoExistente = repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Refeição não encontrada"));

        validarCamposObrigatorios(dto);

        Usuario usuario = usuarioRepository.findById(dto.getUsuario().getId())
                .orElseThrow(() -> new RuntimeException("Usuário não encontrado"));

        refeicaoExistente.setTipo(dto.getTipo());
        refeicaoExistente.setDescricao(dto.getDescricao());
        refeicaoExistente.setData(dto.getData());
        refeicaoExistente.setHorario(dto.getHorario());
        refeicaoExistente.setNivelFome(dto.getNivelFome());
        refeicaoExistente.setNivelSaciedade(dto.getNivelSaciedade());
        refeicaoExistente.setCompanhia(dto.getCompanhia());
        refeicaoExistente.setDistracoes(dto.getDistracoes());
        refeicaoExistente.setEmocoesAntes(dto.getEmocoesAntes());
        refeicaoExistente.setEmocoesDepois(dto.getEmocoesDepois());
        refeicaoExistente.setUsuario(usuario);

        Refeicao atualizada = repository.save(refeicaoExistente);

        return toDTO(atualizada);
    }

    // ===================== Histórico por mês =====================
    public List<RefeicaoDto> historicoPorMes(Integer usuarioId, int mes, int ano) {
        List<Refeicao> lista = repository.findByUsuarioAndMes(usuarioId, mes, ano);
        return lista.stream()
                .map(this::toDTO)
                .collect(Collectors.toList());
    }

    // ===================== Delete =====================
    public void deletarRefeicao(Integer idRefeicao) {
        Refeicao refeicao = repository.findById(idRefeicao)
                .orElseThrow(() -> new RuntimeException("Refeição não encontrada"));
        repository.delete(refeicao);
    }

    // ===================== Conversão DTO ↔ Entity =====================
    private RefeicaoDto toDTO(Refeicao refeicao) {
        if (refeicao == null)
            return null;

        Usuario usuario = refeicao.getUsuario();
        UsuarioDto usuarioDTO = null;
        if (usuario != null) {
            usuarioDTO = new UsuarioDto();
            usuarioDTO.setId(usuario.getId());
            usuarioDTO.setNome(usuario.getNome());
            usuarioDTO.setEmail(usuario.getEmail());
        }

        return new RefeicaoDto(
                refeicao.getIdRefeicao(),
                refeicao.getData(),
                refeicao.getHorario(),
                refeicao.getTipo(),
                refeicao.getDescricao(),
                refeicao.getNivelFome(),
                refeicao.getCompanhia(),
                refeicao.getDistracoes(),
                refeicao.getEmocoesAntes(),
                refeicao.getEmocoesDepois(),
                refeicao.getNivelSaciedade(),
                usuarioDTO);
    }

    private Refeicao toEntity(RefeicaoDto dto, Usuario usuario) {
        return Refeicao.builder()
                .idRefeicao(dto.getIdRefeicao())
                .tipo(dto.getTipo())
                .descricao(dto.getDescricao())
                .data(dto.getData())
                .horario(dto.getHorario())
                .nivelFome(dto.getNivelFome())
                .nivelSaciedade(dto.getNivelSaciedade())
                .companhia(dto.getCompanhia())
                .distracoes(dto.getDistracoes())
                .emocoesAntes(dto.getEmocoesAntes())
                .emocoesDepois(dto.getEmocoesDepois())
                .usuario(usuario)
                .build();
    }

    // ===================== Validação de Campos =====================
    private void validarCamposObrigatorios(RefeicaoDto dto) {
        if (dto.getUsuario() == null || dto.getUsuario().getId() == null) {
            throw new IllegalArgumentException("Usuário inválido");
        }
        if (dto.getData() == null) {
            throw new IllegalArgumentException("Você deve informar a data da refeição");
        }
        if (dto.getHorario() == null) {
            throw new IllegalArgumentException("Você deve informar o horário da refeição");
        }
        if (dto.getTipo() == null || dto.getTipo().isEmpty()) {
            throw new IllegalArgumentException("Você deve informar o tipo da refeição");
        }
        if (dto.getNivelFome() == null) {
            throw new IllegalArgumentException("Você deve informar o nível de fome");
        }
        if (dto.getNivelSaciedade() == null) {
            throw new IllegalArgumentException("Você deve informar o nível de saciedade");
        }
    }
}
